Make sure to run a service worker before firing its activate event

youennf opened the pull request:
#### ff4eced3ff239c434ef397c0c059a8a027a54db5
<pre>
Make sure to run a service worker before firing its activate event
<a href="https://bugs.webkit.org/show_bug.cgi?id=256893">https://bugs.webkit.org/show_bug.cgi?id=256893</a>
rdar://109411104

Reviewed by Ben Nham.

As per spec, we should run the service worker before firing the activate event.
We were not doing so and there might be cases where the service worker will no longer be running between installed and activating steps.
This happens for instance in case the process is terminated due to memory issues.
We align the implementation with the spec by renaming fireActivateEvent to runServiceWorkerAndFireActivateEvent.
Internally, it calls runServiceWorkerIfNecessary before firing the event.
Small refactoring to allow either running a service worker from its object or from its ID.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/service-workers/changing-service-worker-lengthy-worker.py: Added.
(main):
* LayoutTests/http/wpt/service-workers/service-worker-process-crashing-between-installed-and-activating-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/service-worker-process-crashing-between-installed-and-activating.html: Added.
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::runServiceWorkerIfNecessary):
(WebCore::SWServer::runServiceWorker):
(WebCore::SWServer::runServiceWorkerAndFireActivateEvent):
(WebCore::SWServer::fireActivateEvent): Deleted.
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerRegistration.cpp:
(WebCore::SWServerRegistration::activate):

Canonical link: <a href="https://commits.webkit.org/264242@main">https://commits.webkit.org/264242@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4071ecc80a97dda8f348338472018ebbc5f080d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6899 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8498 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7146 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6904 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8422 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7068 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10046 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7021 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7727 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8598 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6260 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14053 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6790 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6347 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9159 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6831 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5596 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6207 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6174 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1688 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10392 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6585 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham reviewed (approved):
Seems reasonable to me.

youennf committed:
Make sure to run a service worker before firing its activate event
https://bugs.webkit.org/show_bug.cgi?id=256893
rdar://109411104

Reviewed by Ben Nham.

As per spec, we should run the service worker before firing the activate event.
We were not doing so and there might be cases where the service worker will no longer be running between installed and activating steps.
This happens for instance in case the process is terminated due to memory issues.
We align the implementation with the spec by renaming fireActivateEvent to runServiceWorkerAndFireActivateEvent.
Internally, it calls runServiceWorkerIfNecessary before firing the event.
Small refactoring to allow either running a service worker from its object or from its ID.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/service-workers/changing-service-worker-lengthy-worker.py: Added.
(main):
* LayoutTests/http/wpt/service-workers/service-worker-process-crashing-between-installed-and-activating-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/service-worker-process-crashing-between-installed-and-activating.html: Added.
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::runServiceWorkerIfNecessary):
(WebCore::SWServer::runServiceWorker):
(WebCore::SWServer::runServiceWorkerAndFireActivateEvent):
(WebCore::SWServer::fireActivateEvent): Deleted.
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerRegistration.cpp:
(WebCore::SWServerRegistration::activate):

Canonical link: https://commits.webkit.org/264242@main
